nfs: design nfs quota (#6112)
rook/rook: design/nfs/nfs-quota.md

Currently, when the user creates NFS PersistentVolumes from an NFS Rook share/export via PersistentVolumeClaim, the provisioner does not provide the specific capacity as requested. For example the users create NFS PersistentVolumes via PersistentVolumeClaim as following:

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: rook-nfs-pv-claim
spec:
  storageClassName: "rook-nfs-share"
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1Mi
```

The client still can use the higher capacity than `1mi` as requested.

This proposal is to add features which the Rook NFS Provisioner will provide the specific capacity as requested from `.spec.resources.requests.storage` field in PersistentVolumeClaim.

## Implementation

The implementation will be use `Project Quota` on xfs filesystem. When the users need to use the quota feature they should use xfs filesystem with `prjquota/pquota` mount options for underlying volume. Users can specify filesystem type and mount options through StorageClass that will be used for underlying volume. For example:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: standard-xfs
parameters:

>   fsType: xfs

Reviewer: just looking at this storageClass, i was thinking that saying `xfs` implied it would result in an underlying volume that it tied to a single host node.  But many types of distributed storage providers (e.g. google PD or amazon AWS) support XFS too, so using XFS in the underlying volume does not mean that the NFS server is tied to a single host. Correct?

is that worth adding a note about here (that many distributed volume types support XFS)?

...
```

If the above conditions are met then the Rook NFS Provisioner will create projects and set the quota limit using [xfs_quota](https://linux.die.net/man/8/xfs_quota) before creating PersistentVolumes based on `.spec.resources.requests.storage` field in PersistentVolumeClaim. Otherwise the Rook NFS Provisioner will provision a PersistentVolumes without creating setting the quota.

To creating the project, Rook NFS Provisioner will invoke the following command

> xfs_quota -x -c project -s -p '*nfs_pv_directory* *project_id*' *projects_file*

And setting quota with the command

> xfs_quota -x -c 'limit -p bhard=*size* *project_id*' *projects_file*

which

1. *nfs_pv_directory* is sub-directory from exported directory that used for NFS PV.
1. *project_id* is unique id `uint16` 1 to 65535.
1. *size* is size of quota as requested.
1. *projects_file* is file that contains *project quota block* for persisting quota state purpose. In case the Rook NFS Provisioner pod is killed, Rook NFS Provisioner pod will restore the quota state based on *project quota block* entries in *projects_file* at startup.
1. *project quota block* is combine of *project_id*:*nfs_pv_directory*:*size*

> Since Rook NFS has possibility to create more than one of nfs share/export that have different underlying volume directory, the *projects_file* will saved on each underlying volume directory. So each nfs share/export will have different *projects_file* and *project_file* will still persist. The *projects_file* will ony created if underlying volume directory mounted as xfs with prjquota mount options. This mean the existence of *project_file* will indicate if quota was enabled. The hierarchy of directory will look like

Reviewer: ```suggestion
Since Rook NFS has the ability to create more than one NFS share/export that have different underlying volume directories, the *projects_file* will be saved on each underlying volume directory. So each NFS share/export will have different *projects_file* and each *project_file* will be persisted. The *projects_file* will only be created if underlying volume directory is mounted as `xfs` with `prjquota` mount options. This mean the existence of *project_file* will indicate if quota was enabled. The hierarchy of directory will look like:
```

> ```console

Reviewer: this tree is super useful to understand the different cases and how quota will be set (or not set) and tracked for each!

Author: @jbw976 Do we still need some diagram to imagine this schema as we discussed in the last couple of weeks? or it will be enough by this tree?

Reviewer: This tree is sufficient, no other diagram is needed :)

/
├── underlying-volume-A (export A) (mounted as xfs with prjquota mount options)
│   ├── projects_file
│   ├── nfs-pv-a (PV-A) (which quota created for)
│   │   ├── data (from PV-A)
│   └── nfs-pv-b (PV-B) (which quota created for)
│       └── data (from PV-B)
├── underlying-volume-B (export B) (mounted as xfs with prjquota mount options)
│   ├── projects_file
│   └── nfs-pv-c (PV-C) (which quota created for)
└── underlying-volume-C (export C) (not mounted as xfs)
    └── nfs-pv-d (PV-D) (quota not created)
```

The hierarchy above is example Rook NFS has 3 nfs share/exports (A, B and C). *project_file* inside underlying-volume-A will contains *project quota block* like

```
1:/underlying-volume-A/nfs-pv-a:size
2:/underlying-volume-A/nfs-pv-b:size
```

*project_file* inside underlying-volume-B will look like

```

> 3:/underlying-volume-B/nfs-pv-c:size

Reviewer: does this need to be 3 (unique and not conflicting with the previous 2 projects?  This is on a separate underlying volume, so we could reuse 1 or 2 here right? I'm hoping we don't have to manage a unique set of project IDs across underlying volumes

Author: Yes, you are right. We should have a different set of project IDs across underlying volumes

# NFS Quota

## Background

> Currently, when the user creates PersistentVolumes from an NFS Rook share/export via PersistentVolumeClaim, the provisioner does not provide the spesific capacity as requested. For example the users create PersistentVolumeClaim as following:

Reviewer 2: typo: specific

Author: whoops, thanks 😁

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: rook-nfs-pv-claim
spec:
  storageClassName: "rook-nfs-share"
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1Mi
```

The client still can use the higher capacity than `1mi` as requested.

This proposal is to add features which the Rook NFS Provisioner will provide the specific capacity as requested from `.spec.resources.requests.storage` field in PersistentVolumeClaim.

## Implementation

The implementation will be use `Project Quota` on xfs filesystem. When the users need to use the quota feature they should use xfs filesystem with `prjquota/pquota` mount options for underlying volume. Users can specify filesystem type and mount options through StorageClass that will be used for underlying volume. For example:

> ```yaml

Reviewer: ah, this is the *underlying* volume that is underneath the NFS share/export, NOT the storageClass that is used for a NFS based PVC.  I was confused at first, even though you did mention "underlying volume" in the sentence before. 

this may be helped from more clarity (or a little diagram), of the layering of volumes that are in play here
* NFS PV
* NFS share/export
* underlying volume

apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: standard-xfs
parameters:
  fsType: xfs
mountOptions:
  - prjquota
...
```

Then the underlying PersistentVolumeClaim should be used that StorageClass 

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: nfs-default-claim
spec:
  storageClassName: "standard-xfs"
  accessModes:
  - ReadWriteOnce
...
```

> If the above conditions are met then the Rook NFS Provisioner will create projects and set the quota limit using [xfs_quota](https://linux.die.net/man/8/xfs_quota) before creating PersistentVolumes based on `.spec.resources.requests.storage` field in PersistentVolumeClaim. Otherwise the Rook NFS Provisioner will provision a PersistentVolumes without creating setting the quota.

Reviewer: what does it look like to "create projects and set the quota limit", can you add an example command that we will call?

do we need to update the container image to now also include this `xfs_quota` binary?

Author: yeah, I'll add the example command for creating projects and setting the quota limit that will be invoked by nfs-provisioner.

yes, we need to update the container image to include the `xfs_quota` binary. will update it on the doc as well

> If the above conditions are met then the Rook NFS Provisioner will create projects and set the quota limit using [xfs_quota](https://linux.die.net/man/8/xfs_quota) before creating PersistentVolumes based on `.spec.resources.requests.storage` field in PersistentVolumeClaim. Otherwise the Rook NFS Provisioner will provision a PersistentVolumes without creating setting the quota.

Author: Who has the responsibility for setting a quota for requested PV is  `nfs-provisioner`. For underlying PV, Kubelet has a responsibility to mount the PV into the pod depend on `mountOptions` and `parameters.fsType` on storageclass.

Yup, I'll make this design more detail :bowing_man:

> If the above conditions are met then the Rook NFS Provisioner will create projects and set the quota limit using [xfs_quota](https://linux.die.net/man/8/xfs_quota) before creating PersistentVolumes based on `.spec.resources.requests.storage` field in PersistentVolumeClaim. Otherwise the Rook NFS Provisioner will provision a PersistentVolumes without creating setting the quota.

Author: Yes, they are different. each PVC which backed by NFS export will be a different `projects`. let's say PVC A will have projects id 1 and PVC B will have projects id 2 and so on. And each project has different size/capacity based on `.spec.resources.requests.storage`.

The mapping for PVC and `projects` will look like
|PVC|project entries|
|-|-|
|A|1:/pv-a-path:size|
|B|2:/pv-b-path:size|

The nfs-provisioner will also create a `projects` file for saving those project entries, so it will be persistent in case nfs-provisioner restarted/killed then nfs-provisioner can restore the state of quota for each directory (PV which backed by NFS export).

> If the above conditions are met then the Rook NFS Provisioner will create projects and set the quota limit using [xfs_quota](https://linux.die.net/man/8/xfs_quota) before creating PersistentVolumes based on `.spec.resources.requests.storage` field in PersistentVolumeClaim. Otherwise the Rook NFS Provisioner will provision a PersistentVolumes without creating setting the quota.

Reviewer: Thank you for clarifying.  where would this `projects` file be persisted to exactly?

> ### Why XFS

Author: I haven't tested yet, but it should work as well with condition `fstype` is xfs and with `prjquota` mount options in rbd storageclass.

https://github.com/rook/rook/blob/24b0bdd119cf5cf453865af91bc81fe0ac88f5ff/cluster/examples/kubernetes/ceph/csi/rbd/storageclass.yaml#L53

> ### Why XFS

Reviewer: This is good to know, that Ceph underlying volumes can support this as well.  It may be *nice* to include this in the instructions, but it is definitely not a requirement to do so :)
